### DISCUSSION TITLE
Help implementing perspective with Mapim

### DISCUSSION THREAD

**Discussion by jsavell** [libvips/libvips#3859] [Q&A]
I've been working on implementing a rotation/perspective functionality in netvips that would output something like this, dependent on user input for a rotation value.

![skew-ex](https://github.com/kleisauke/net-vips/assets/5373296/7c3c5d72-c835-4fcc-b011-e9b8be145504)

Using all of the below resources, I've managed to get something close to the above image, better for some rotational values than others.

But it's just not right.

This is what I'm doing, but it's really just me trialing and erroring. I just haven't been able to get to a solid understanding of the math involved, how to calculate 'tie points', how exactly netvips/libvips is using the mapim parameters I'm feeding it, etc.

```
            using (Image i = Image.Xyz(image.Width, image.Height))
            {
                var width = image.Width;
                var height = image.Height;
                double tRot = -(rotation*0.01);//keep this

                var p0 = (rotation > 0) ? 1-(tRot):1-1*(-tRot*22); //1.5+tRot;
                var p1 = 0;
                var p2 = (rotation > 0) ? 0:1;
                var p3 = (rotation > 0) ? -.111:.111;
                var p4 = (rotation > 0) ? 1:1.2;
                var p5 = (rotation > 0) ? 0:-45;
                var p6 = (rotation > 0) ? -.0005:.0005;
                var p7 = 0;
                double[] T = new double[] {p0, p1, p2, p3, p4, p5, p6, p7};


                var x = (i[0] * T[0] + i[1] * T[1] + T[2]) / (i[0] * T[6] + i[1] * T[7] + 1);
                var y = (i[0] * T[3] + i[1] * T[4] + T[5]) / (i[0] * T[6] + i[1] * T[7] + 1);
                var m = x.Bandjoin(y);

                return image.Mapim(m);
            }

```
Again, the code is functioning to transform the image, I'm just struggling to tell netvips/libvips how I want the image to be transformed. 

I'd appreciate any help or guidance to get the output closer to the above example images. Thanks.

https://stackoverflow.com/questions/50202733/how-to-perform-perspective-distort-transformation-in-vips/50219715#50219715
https://github.com/libvips/libvips/discussions/3140
https://github.com/opencv/opencv/blob/8f0edf6a1c7015dd76abb7dc8a82ea36d5500a72/modules/imgproc/src/imgwarp.cpp#L3383
https://stackoverflow.com/questions/65289489/how-to-calculate-a-transform-matrix-from-set-of-tie-points-for-perspective-trans
https://math.stackexchange.com/questions/296794/finding-the-transform-matrix-from-4-projected-points-with-javascript
https://franklinta.com/2014/09/08/computing-css-matrix3d-transforms/

**Comment by philocalyst** [comment]
Any luck @jsavell ? There's an imagemagick script that is capable of this that I would love to see someone expand to libvips. Way beyond my capabilities, but check it out -- there's even an entire writeup of the math behind it.

http://www.fmwconcepts.com/imagemagick/3Drotate/index.php
http://www.fmwconcepts.com/imagemagick/3Drotate/3DROTATE.pdf


Imagemagick and libvips are not similar at all but most functions have an alternative if you look hard enough. You seem like you're willing to look.

**Reply by philocalyst** [reply]
P.S
If you do to pursue it using the script as a resource, please respect the license and let the original creator know!
